tentacle/src/transports/tls.rs

```diff
@@ -0,0 +1,254 @@
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use crate::runtime::TcpListener;
+use crate::service::TlsConfig;
+use crate::transports::{find_type, TransportFuture, TransportType};
+use crate::{
+    error::TransportErrorKind,
+    multiaddr::{Multiaddr, Protocol},
+    session::AsyncRw,
+    transports::{tcp_dial, tcp_listen, Transport},
+    utils::{dns::DnsResolver, multiaddr_to_socketaddr, socketaddr_to_multiaddr},
+};
+use futures::channel::mpsc::{channel, Receiver, Sender};
+use std::sync::Arc;
+use tokio::io;
+use tokio_rustls::rustls::{ClientConfig, ServerConfig};
+use tokio_rustls::webpki::DNSNameRef;
+use tokio_rustls::{TlsAcceptor, TlsConnector};
+
+pub type TlsStream = Box<dyn AsyncRw + Send + Unpin + 'static>;
+
+/// Tls listen bind
+async fn bind(
+    address: impl Future<Output = Result<Multiaddr>>,
+    timeout: Duration,
+    reuse: bool,
+    tls_config: ServerConfig,
+) -> Result<(Multiaddr, TlsListener)> {
+    let addr = address.await?;
+    match multiaddr_to_socketaddr(&addr) {
+        Some(socket_address) => {
+            let (local_addr, tcp) = tcp_listen(socket_address, reuse).await?;
+
+            let mut listen_addr = socketaddr_to_multiaddr(local_addr);
+
+            if let TransportType::Tls(s) = find_type(&addr) {
+                listen_addr.push(Protocol::Tls(Cow::Borrowed(&s)));
+                Ok((listen_addr, TlsListener::new(timeout, tcp, tls_config)))
+            } else {
+                Err(TransportErrorKind::NotSupported(addr))
+            }
+        }
+        None => Err(TransportErrorKind::NotSupported(addr)),
+    }
+}
+
+/// Tls connect
+async fn connect(
+    address: impl Future<Output = Result<Multiaddr>>,
+    timeout: Duration,
+    original: Option<Multiaddr>,
+    bind_addr: Option<SocketAddr>,
+    tls_client_config: ClientConfig,
+    domain_name: String,
+) -> Result<(Multiaddr, TlsStream)> {
+    let addr = address.await?;
+    let connector = TlsConnector::from(Arc::new(tls_client_config));
+    match multiaddr_to_socketaddr(&addr) {
+        Some(socket_address) => {
+            let stream = tcp_dial(socket_address, bind_addr, timeout).await?;
+
+            let domain_name = DNSNameRef::try_from_ascii_str(&domain_name)
+                .map_err(|_| TransportErrorKind::TlsError("invalid dnsname".to_string()))?;
+            Ok((
+                original.unwrap_or(addr),
+                Box::new(
+                    connector
+                        .connect(domain_name, stream)
+                        .await
+                        .map_err(TransportErrorKind::Io)?,
+                ),
+            ))
+        }
+        None => Err(TransportErrorKind::NotSupported(original.unwrap_or(addr))),
+    }
+}
+
+pub struct TlsListener {
+    inner: TcpListener,
+    timeout: Duration,
+    sender: Sender<(Multiaddr, TlsStream)>,
+    pending_stream: Receiver<(Multiaddr, TlsStream)>,
+    tls_config: Box<ServerConfig>,
```

Comment:
if I understand correctly, this can be changed to:
```
tls_config: Arc<ServerConfig>
```

and can use Arc#clone in acceptor:
```
let acceptor = TlsAcceptor::from(Arc::clone(&self.tls_config));
```